Lib/exception_group.py

```diff
@@ -0,0 +1,215 @@
+            return None
+        result = []
+        e = self.subgroup([exc])
+        while e:
+            tb = e.__traceback__
+            while tb:
+                result.append(tb.tb_frame)
+                tb = tb.tb_next
+            if isinstance(e, ExceptionGroup):
+                assert len(e.excs) == 1 and exc in e
+                e = e.excs[0]
+            else:
+                assert e is exc
+                e = None
+        return result
+
+    def push_frame(self, frame):
+        import types
+        self.__traceback__ = types.TracebackType(
+            self.__traceback__, frame, 0, 0)
+
+    @staticmethod
+    def render(exc, tb=None, indent=0):
+        output = []
+        output.append(f"{exc}")
+        tb = tb or exc.__traceback__
+        while tb and not isinstance(tb, TracebackGroup):
+            output.append(f"{' '*indent} {tb.tb_frame}")
+            tb = tb.tb_next
+        if isinstance(exc, ExceptionGroup):
+            tbg = exc.__traceback_group__
+            assert isinstance(tbg, TracebackGroup)
+            indent += 4
+            for e in exc.excs:
+                t = tbg.tb_next_map[id(e)]
+                output.append('---------------------------------------')
+                output.extend(ExceptionGroup.render(e, t, indent))
+        for l in output:
+            print(l)
+        return output
+
+    def __iter__(self):
+        ''' iterate over the individual exceptions (flattens the tree) '''
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup):
+                for e_ in e:
+                    yield e_
+            else:
+                yield e
+
+    def __len__(self):
+        l = 0
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup):
+                l += len(e)
+            else:
+                l += 1
+        return l
+
+    def __repr__(self):
+        return f"ExceptionGroup({self.excs})"
+
+    @staticmethod
+    def catch(types, handler):
+        return ExceptionGroupCatcher(types, handler)
+
+class ExceptionGroupCatcher:
+    """ Based on trio.MultiErrorCatcher """
+
+    def __init__(self, types, handler):
+        """ Context manager to catch and handle ExceptionGroups
+
+        types: the exception types that this handler is interested in
+        handler: a function that takes an ExceptionGroup of the
+        matched type and does something with them
+
+        Any rest exceptions are raised at the end as another
+        exception group
+        """
+        self.types = types
+        self.handler = handler
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, etype, exc, tb):
+        if exc is not None and isinstance(exc, ExceptionGroup):
+            match, rest = exc.split(self.types)
+
+            if not match:
+                # Let the interpreter reraise the exception
+                return False
+
+            handler_excs = self.handler(match)
+            if handler_excs == match:
```

Comment:
Um, EG doesn’t define ‘__eq__’ so this uses ‘is’. Maybe that’s what we want, then write it that way?